scrapy_po/utils.py

```diff
@@ -1,7 +1,122 @@
 # -*- coding: utf-8 -*-
+import inspect
+from typing import Tuple, Dict, Type, Callable
+
+import andi
+from scrapy.http import Response
+from scrapy.utils.defer import maybeDeferred_coro
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+from scrapy_po.webpage import Injectable
+from scrapy_po.page_input_providers import providers
+
 
 def get_callback(request, spider):
-    """ Get request.callback of a scrapy.Request, as a callable """
+    """Get request.callback of a scrapy.Request, as a callable."""
     if request.callback is None:
         return getattr(spider, 'parse')
     return request.callback
+
+
+class DummyResponse(Response):
+
+    def __init__(self, url, request=None):
+        super(DummyResponse, self).__init__(url=url, request=request)
+
+
+def is_callback_using_response(callback: Callable):
+    """Check whether the request's callback method is going to use response."""
+    spec = inspect.getfullargspec(callback)
+    try:
+        arg_name = spec.args[1]  # first index is self, second is response
```

Comment:
Just to be sure, could you please check (add a test?) how it works with e.g. partial callbacks:

```py

from functools import partial

    def start_requests(self):
        yield Request(partial(self.parse, use_x=True))

    def parse(self, response, use_x):
        # ...
```

Users should be using cb_kwargs, but partial or lambdas may be used in a real code.

Reply:
Well, it seems like Andi does not support partials:

```
Traceback (most recent call last):
  File "/Users/victortorres/.virtualenvs/scrapy-po/lib/python3.7/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/Users/victortorres/.virtualenvs/scrapy-po/lib/python3.7/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/Users/victortorres/.virtualenvs/scrapy-po/lib/python3.7/site-packages/scrapy/core/downloader/middleware.py", line 51, in process_response
    response = yield deferred_from_coro(method(request=request, response=response, spider=spider))
  File "/Users/victortorres/.virtualenvs/scrapy-po/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/Users/victortorres/git/shub/scrapy-po/scrapy_po/middleware.py", line 30, in process_response
    plan, provider_instances = build_plan(callback, response)
  File "/Users/victortorres/git/shub/scrapy-po/scrapy_po/middleware.py", line 53, in build_plan
    externally_provided=provider_instances.keys()
  File "/Users/victortorres/.virtualenvs/scrapy-po/lib/python3.7/site-packages/andi/andi.py", line 264, in plan
    dependency_stack=None)
  File "/Users/victortorres/.virtualenvs/scrapy-po/lib/python3.7/site-packages/andi/andi.py", line 294, in _plan
    arguments = inspect(class_or_func)
  File "/Users/victortorres/.virtualenvs/scrapy-po/lib/python3.7/site-packages/andi/andi.py", line 24, in inspect
    annotations = get_type_hints(func, globalns)
  File "/Users/victortorres/miniconda3/lib/python3.7/typing.py", line 993, in get_type_hints
    'or function.'.format(obj))
TypeError: functools.partial(<bound method PartialCallbackSpider.parse of <PartialCallbackSpider 'test_spider' at 0x10f487dd8>>, use_x=True) is not a module, class, method, or function.
```

Reply:
could you please raise an issue in andi repo?

Reply:
Done https://github.com/scrapinghub/andi/issues/16

Reply:
Thanks @victor-torres!

What about lambdas?

Reply:
It seems like it's not possible to extract type annotations from lambda functions. 

I have tried to annotate them this way (found an example on StackOverflow) but it doesn't seem to work.
```
class LambdaSkipDownloadSpider(scrapy.Spider):

    url = None

    def start_requests(self):
        cb: Callable[[DummyResponse], Dict] = lambda response: {'response': response}
        yield Request(url=self.url, callback=cb)
```

Inspect results:

```
(Pdb) spec
FullArgSpec(args=['response'], varargs=None, varkw=None, defaults=None, kwonlyargs=[], kwonlydefaults=None, annotations={})
(Pdb) callback
<function LambdaSkipDownloadSpider.start_requests.<locals>.<lambda> at 0x108da3bf8>
```

Any suggestion?

Reply:
Aha, a good call. Let's document somewhere that lambdas don't work, and won't work - not sure where.

Reply:
I've put something in the README.rst file.